src/Entity/SmsMessage.php

```diff
@@ -0,0 +1,411 @@
+    $this->set('recipient_entity', $entity);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQueued() {
+    return $this->get('queued')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setQueued($is_queued) {
+    $this->set('queued', $is_queued);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCreatedTime() {
+    return $this->get('created')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSendTime() {
+    return $this->get('send_on')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setSendTime($send_time) {
+    $this->set('send_on', $send_time);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getProcessedTime() {
+    return $this->get('processed')->value;
+  }
+  /**
+   * {@inheritdoc}
+   */
+  public function setProcessedTime($processed) {
+    $this->set('processed', $processed);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
+    // Identifiers.
+    $fields['id'] = BaseFieldDefinition::create('integer')
+      ->setLabel(t('SMS message ID'))
+      ->setDescription(t('The SMS message ID.'))
+      ->setReadOnly(TRUE)
+      ->setSetting('unsigned', TRUE);
+
+    $fields['uuid'] = BaseFieldDefinition::create('uuid')
+      ->setLabel(t('UUID'))
+      ->setDescription(t('The SMS message UUID.'))
+      ->setReadOnly(TRUE);
+
+    $fields['gateway'] = BaseFieldDefinition::create('entity_reference')
+      ->setLabel(t('Gateway Plugin'))
+      ->setDescription(t('The gateway plugin instance.'))
+      ->setSetting('target_type', 'sms_gateway')
+      ->setReadOnly(TRUE)
+      ->setRequired(TRUE);
+
+    $fields['direction'] = BaseFieldDefinition::create('integer')
+      ->setLabel(t('Transmission direction'))
+      ->setDescription(t('Transmission direction, See SmsMessageInterface::DIRECTION_*.'))
+      ->setReadOnly(TRUE)
+      ->setSetting('unsigned', FALSE)
+      ->setSetting('size', 'tiny');
+
+    // Sender and receivers.
+    $fields['sender'] = BaseFieldDefinition::create('telephone')
+      ->setLabel(t('Sender phone number'))
+      ->setDescription(t('The phone number of the sender.'))
+      ->setDefaultValue('')
+      ->setRequired(FALSE);
+
+    $fields['sender_entity'] = BaseFieldDefinition::create('dynamic_entity_reference')
+      ->setLabel(t('Sender entity'))
+      ->setDescription(t('The entity who sent the SMS message.'))
+      ->setRequired(FALSE);
+
+    $fields['recipient'] = BaseFieldDefinition::create('telephone')
+      ->setLabel(t('Sender phone number'))
+      ->setDescription(t('The phone number of the sender.'))
+      ->setDefaultValue('')
+      ->setRequired(FALSE)
+      ->setCardinality(BaseFieldDefinition::CARDINALITY_UNLIMITED);
+
+    $fields['recipient_entity'] = BaseFieldDefinition::create('dynamic_entity_reference')
```

Comment:
Of course, the comment on the `sender` and `sender_entity` also applies here too. We could just store the phone numbers instead of the whole entity, then get the entities, if we need to, using phone number verification storage.